index.js

```diff
@@ -1,105 +1,160 @@
-import isFullwidthCodePoint from 'is-fullwidth-code-point';
 import ansiStyles from 'ansi-styles';
+import isFullwidthCodePoint from 'is-fullwidth-code-point';
 
-const astralRegex = /^[\uD800-\uDBFF][\uDC00-\uDFFF]$/;
+const ESCAPES = new Set([27, 155]);
 
-const ESCAPES = [
-	'\u001B',
-	'\u009B'
-];
+const endCodesSet = new Set();
+const endCodesMap = new Map();
+for (const [start, end] of ansiStyles.codes) {
+	endCodesSet.add(ansiStyles.color.ansi(end));
+	endCodesMap.set(ansiStyles.color.ansi(start), ansiStyles.color.ansi(end));
+}
 
-const wrapAnsi = code => `${ESCAPES[0]}[${code}m`;
+function getEndCode(code) {
+	if (endCodesSet.has(code)) {
+		return code;
+	}
 
-const checkAnsi = (ansiCodes, isEscapes, endAnsiCode) => {
-	let output = [];
-	ansiCodes = [...ansiCodes];
+	if (endCodesMap.has(code)) {
+		return endCodesMap.get(code);
+	}
 
-	for (let ansiCode of ansiCodes) {
-		const ansiCodeOrigin = ansiCode;
-		if (ansiCode.includes(';')) {
-			ansiCode = ansiCode.split(';')[0][0] + '0';
-		}
+	code = code.slice(2);
+	if (code.includes(';')) {
+		code = code[0] + '0';
+	}
 
-		const item = ansiStyles.codes.get(Number.parseInt(ansiCode, 10));
-		if (item) {
-			const indexEscape = ansiCodes.indexOf(item.toString());
-			if (indexEscape === -1) {
-				output.push(wrapAnsi(isEscapes ? item : ansiCodeOrigin));
-			} else {
-				ansiCodes.splice(indexEscape, 1);
-			}
-		} else if (isEscapes) {
-			output.push(wrapAnsi(0));
-			break;
-		} else {
-			output.push(wrapAnsi(ansiCodeOrigin));
-		}
+	const returnValue = ansiStyles.codes.get(Number.parseInt(code, 10));
+	if (returnValue) {
+		return ansiStyles.color.ansi(returnValue);
 	}
 
-	if (isEscapes) {
-		output = output.filter((element, index) => output.indexOf(element) === index);
+	return ansiStyles.reset.open;
+}
 
-		if (endAnsiCode !== undefined) {
-			const fistEscapeCode = wrapAnsi(ansiStyles.codes.get(Number.parseInt(endAnsiCode, 10)));
-			// TODO: Remove the use of `.reduce` here.
-			// eslint-disable-next-line unicorn/no-array-reduce
-			output = output.reduce((current, next) => next === fistEscapeCode ? [next, ...current] : [...current, next], []);
+function findNumberIndex(string) {
+	for (let index = 0; index < string.length; index++) {
+		const charCode = string.charCodeAt(index);
+		if (charCode >= 48 && charCode <= 57) {
```

Comment:
Magic numbers. Either add a comment about them or move them to be constants with a descriptive name.